src/tokenizedSubaccounts/PPTSA.sol

```diff
@@ -0,0 +1,597 @@
+  /// this logic is so the vault can execute as taker and maker while keeping the code aware of all trades
+  function _verifyRfqAction(IMatching.Action memory action, bytes memory extraData) internal view {
+    uint maxFee;
+    IRfqModule.TradeData[] memory makerTrades;
+    // non 0 length means the executor is a taker.
+    // Since taker does not have an array of the trades, we require the executor to pass in the trades as extraData.
+    // If the extraData is 0, we know the executor is a maker so the trades are included in the action.data
+    if (extraData.length == 0) {
+      IRfqModule.RfqOrder memory makerOrder = abi.decode(action.data, (IRfqModule.RfqOrder));
+      maxFee = makerOrder.maxFee;
+      makerTrades = makerOrder.trades;
+    } else {
+      IRfqModule.TakerOrder memory takerOrder = abi.decode(action.data, (IRfqModule.TakerOrder));
+      if (keccak256(extraData) != takerOrder.orderHash) {
+        revert PPT_TradeDataDoesNotMatchOrderHash();
+      }
+      maxFee = takerOrder.maxFee;
+      makerTrades = abi.decode(extraData, (IRfqModule.TradeData[]));
+    }
+
+    (StrikeData memory lowerStrike, StrikeData memory higherStrike) = _validateAndSortStrikes(makerTrades);
+    _verifyHigherStrikeAmount(higherStrike.tradeAmount, extraData.length != 0);
+
+    _verifyRfqExecute(lowerStrike, higherStrike, maxFee);
+  }
+
+  /// @dev the following function validates whether the maker is selling or buying a spread
+  /// The following table determines whether or not the high strike should be sold or bought
+  ///
+  /// isCallSpread | isLongSpread | isTaker | makerShouldBeSellingSpread
+  ///      0       |       0      |    0    |   0 (higherStrike should be < 0)
+  ///      0       |       0      |    1    |   1 (higherStrike should be > 0)
+  ///      0       |       1      |    0    |   1
+  ///      0       |       1      |    1    |   0
+  ///      1       |       0      |    0    |   1
+  ///      1       |       0      |    1    |   0
+  ///      1       |       1      |    0    |   0
+  ///      1       |       1      |    1    |   1
+  function _verifyHigherStrikeAmount(int strikeAmount, bool isTaker) internal pure {
+    PPTSAStorage memory $ = _getPPTSAStorage();
+    bool makerShouldBeSellingSpread = ($.isCallSpread == $.isLongSpread) ? isTaker : !isTaker;
+    if (makerShouldBeSellingSpread && strikeAmount <= 0) {
+      revert PPT_InvalidHighStrikeAmount();
+    } else if (!makerShouldBeSellingSpread && strikeAmount >= 0) {
+      revert PPT_InvalidHighStrikeAmount();
+    }
+  }
+
+  function _validateAndSortStrikes(IRfqModule.TradeData[] memory makerTrades)
+    internal
+    view
+    returns (StrikeData memory lowerStrike, StrikeData memory higherStrike)
+  {
+    if (makerTrades.length != 2) {
+      revert PPT_InvalidParams();
+    }
+    StrikeData memory strike1 = _validateAndCreateIndividualStrike(makerTrades[0]);
+    StrikeData memory strike2 = _validateAndCreateIndividualStrike(makerTrades[1]);
```

Comment:
There's definitely some inefficiency here:
- checking the expiry bounds twice when you check the expiry is the same below
- fetching feed values multiple times for black76

Don't think optimising is that important here at the expense of messy code - so would leave it as is unless you can think of some nice way to move things around

Reply:
I tweaked it a ltittle bit. I ended up keeping the expiry bounds check double counted since it helps preventing some weird errors if the expiry is super off when doing the blackscholes. Keeps errors a tiny bit cleaner IMO